File: KinotirApi/src/main/java/com/kinotir/api/mappers/HtmlFilmDetailMapper.java
package com.kinotir.api.mappers;

import com.kinotir.api.FilmDetailsJson;
import com.kinotir.api.ImageUrlJson;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.List;


class HtmlFilmDetailMapper implements Mapper<String, FilmDetailsJson> {

    private MovieDateMapper dateMapper = new MovieDateMapper();

    @Override
    public FilmDetailsJson map(String html) {
        if (html == null) {
            return null;
        }

        Document doc = Jsoup.parse(html);
        if (doc == null) {
            return null;
        }

        String buyTicketUrl = doc.select("a.buy-ticket-btn").attr("href");

        Element info = doc.select(".film .info").first();
        if (info == null) {
            return null;
        }

        String posterUrl = info.select(">.additional-poster>.image>img").attr("src");

        FilmDetailsJson result = new FilmDetailsJson();
        result.setPosterUrl(posterUrl);
        result.setBuyTicketLink(buyTicketUrl);

        Element mainInfo  = info.select(".main-info").first();
        if (mainInfo != null) {
            parseFeatures(info.select(".features"), result);
            String description = info.select(".description").text();
            result.setDescription(description);
        }

        parseImages(info.select(".slider"), result);
        parseTrailers(info.select(".trailer"), result);

        return result;
    }

    private void parseFeatures(Elements src, FilmDetailsJson destination) {
        src = src.select(">li");

        for (Element el: src) {
            String key = el.select("label").text().trim().toLowerCase();
            String val = el.select("div").text().trim();
            if (key.contains("старт")) {
                destination.setPubDate(dateMapper.map(val));
            } else if (key.contains("страна")) {
                destination.setCountry(val);
            } else if (key.contains("режисер")) {
                destination.setDirector(val);
            } else if (key.contains("сценарий")) {
                destination.setScenario(val);
            } else if (key.contains("в ролях")) {
                destination.setActors(val);
            } else if (key.contains("жанр")) {
                destination.setGenre(val);
            } else if (key.contains("бюджет")) {
                destination.setBudget(val);
            } else if (key.contains("возраст")) {
                destination.setAgeLimit(val);
            } else if (key.contains("продолжительность")) {
                destination.setDuration(val);
            }
        }
    }

    private void parseImages(Elements src, FilmDetailsJson destination) {
        List<ImageUrlJson> urls = new ArrayList<>();
        src = src.select(">a");
        for (Element el: src) {
            String bigImg = el.attr("href");
            String smallImg = el.select("img").attr("src");

            if (smallImg != null && !smallImg.isEmpty()) {
                urls.add(new ImageUrlJson(bigImg, smallImg));
            }
        }

        destination.setImageUrls(urls);
    }

    private void parseTrailers(Elements src, FilmDetailsJson destination) {
        if (src == null) return;
        List<String> trailerUrls = new ArrayList<>();
        src = src.select(">iframe");
        for (Element el: src) {
            String url = el.attr("src");

            if (url != null && !url.isEmpty()) {
                trailerUrls.add(url);
            }
        }

        destination.setTrailerUrls(trailerUrls);
    }
}

File: KinotirApi/src/main/java/com/kinotir/api/mappers/HtmlFilmDetailMapper.kt
package com.kinotir.api.mappers

import com.kinotir.api.FilmDetailsJson
import com.kinotir.api.ImageUrlJson
import org.jsoup.Jsoup
import org.jsoup.select.Elements

internal class HtmlFilmDetailMapper : Mapper<String?, FilmDetailsJson?> {

    private val dateMapper = MovieDateMapper()

    override fun map(src: String?): FilmDetailsJson? {
        val doc = src?.let { Jsoup.parse(it) } ?: return null
        val info = doc.select(".film .info").first() ?: return null

        val result = FilmDetailsJson(
            title = doc.select(".film h1 span")?.text(),
            posterUrl = info.select(">.additional-poster>.image>img")?.attr("src"),
            buyTicketLink = doc.select("a.buy-ticket-btn").attr("href"),
            description = info.select(".description")?.text()
        )
        parseFeatures(info.select(".features"), result)
        parseImages(info.select(".slider"), result)
        parseTrailers(info.select(".trailer"), result)
        return result
    }

    private fun parseFeatures(src: Elements?, destination: FilmDetailsJson) {
        src?.select(">li")
            ?.forEach {
                val key = it.select("label")?.text()?.trim()?.lowercase() ?: ""
                val value = it.select("div")?.text()?.trim() ?: ""
                when {
                    key.contains("старт") -> destination.pubDate = dateMapper.map(value)
                    key.contains("страна") -> destination.country = value
                    key.contains("режисер") -> destination.director = value
                    key.contains("сценарий") -> destination.scenario = value
                    key.contains("в ролях") -> destination.actors = value
                    key.contains("жанр") -> destination.genre = value
                    key.contains("бюджет") -> destination.budget = value
                    key.contains("возраст") -> destination.ageLimit = value
                    key.contains("продолжительность") -> destination.duration = value
                }
            }
    }

    private fun parseImages(src: Elements?, destination: FilmDetailsJson) {
        destination.imageUrls = src?.select(">a")
            ?.mapNotNull {
                val bigImg = it.attr("href") ?: ""
                val smallImg = it.select("img")?.attr("src") ?: ""
                if (smallImg.isBlank()) null
                else ImageUrlJson(bigImg, smallImg)
            }
            ?: emptyList()
    }

    private fun parseTrailers(src: Elements?, destination: FilmDetailsJson) {
        destination.trailerUrls = src?.select(">iframe")
            ?.mapNotNull { it.attr("src") }
            ?.filter { it.isNotBlank() }
            ?: emptyList()
    }
}
